### PR TITLE
Content-Location with QUERY

### LINKED ISSUE #1745: Describe role of Content-Location and Location fields
@reschke mentioned in #1552 that there should be a way for the server to indicate a GET-able equivalent to the query if one exists; this has the benefit of being more easily cacheable by intermediaries which don't understand QUERY.

Based on SEMANTICS, I think the right mechanism for this would be Content-Location.  A server can optionally indicate the GET-able location of the query to which it is responding.

### PR DESCRIPTION
**Pull request by MikeBishop**
Fixes #1745.

### CODE REVIEW

**Comment by mnot** [comment]
`Content-Location` doesn't indicate whether the resource on the other end will return *these* results in perpetuity, or whether it'll return the results to the query (which may change over time). 

So it might be better to be explicit -- if the server wants to create a resource for these results, use `201 Created` with `Location` that explicitly points to a resource that represents this query, and which can change over time. The server can choose whether to include the query results on the 201, of course.

**Comment by asbjornu** [comment]
Could perhaps a `Link` header be used to disambiguate the two use-cases?

```http
Link: </updated-results>; rel="latest-version", </these-results>; rel="memento"
```

The chosen link relations are probably not appropriate, I just picked some that exists to provide a complete example.

**Comment by MikeBishop** [comment]
Per discussion in Brisbane side-meeting:

- Need to add text indicating what it means if the C-L target errors
- Add text clarifying that the resource can be cached despite the different URI (this is cautioned against in SEMANTICS)

File: draft-ietf-httpbis-safe-method-w-body.xml [hunk 1]
```diff
@@ -200,6 +200,14 @@ q=foo&amp;limit=10&amp;sort=-published
       using an HTTP GET request. Various non-normative examples of successful
       QUERY responses are illustrated in <xref target="examples" />.
     </t>
+
```
Comment (reschke) [inline draft-ietf-httpbis-safe-method-w-body.xml:203]:
```suggestion
```

File: draft-ietf-httpbis-safe-method-w-body.xml [hunk 2]
```diff
@@ -200,6 +200,14 @@ q=foo&amp;limit=10&amp;sort=-published
       using an HTTP GET request. Various non-normative examples of successful
       QUERY responses are illustrated in <xref target="examples" />.
     </t>
+    
+    <t>
+      A successful response to a QUERY request &MAY; include a Content-Location
```
Comment (reschke) [inline draft-ietf-httpbis-safe-method-w-body.xml:205]:
```suggestion
      A successful response to a QUERY request can include a Content-Location
```

**Review by MikeBishop — COMMENTED** [review]
Review feedback from @reschke.

File: draft-ietf-httpbis-safe-method-w-body.xml [hunk 3]
```diff
@@ -201,6 +201,17 @@ q=foo&amp;limit=10&amp;sort=-published
       QUERY responses are illustrated in <xref target="examples" />.
     </t>
 
+    <t>
+      A successful response to a QUERY request &MAY; include a Content-Location
+      header field (see <xref target="HTTP" section="8.7"/>) containing an
+      identifier for a specific resource corresponding to the results of the
+      operation. A client can send a GET request for the indicated URI to repeat
```
Comment (martinthomson) [inline draft-ietf-httpbis-safe-method-w-body.xml:208]:
```suggestion
      operation. A client can send a GET request for the indicated URI to repeat
```

So is this "A client can" or "This represents a claim from the server that a client can"

**Review by martinthomson — COMMENTED** [review]
I see nothing here about cacheability or origins.  I know that I left early, but I thought that we had concluded that the server claims about the Content-Location resource would be treated differently based on whether the indicated resource was on the same origin (or not).

**Comment by MikeBishop** [comment]
@martinthomson, I've added language restricting it to same-origin; is there a stronger statement you were expecting to see here?

**Review by reschke — APPROVED** [review]

**Comment by reschke** [comment]
@mnot , @royfielding , @martinthomson - should we go ahead with this?

**Comment by royfielding** [comment]
No.

See https://github.com/httpwg/http-extensions/issues/1745#issuecomment-2130499002

**Comment by reschke** [comment]
IETF 120 feedback: define both (the resource that identfies the result of *this* query), and a resource that can be used with GET to repeat the QUERY itself.

**Comment by reschke** [comment]
This looks good (right, @royfielding ?)

Looking at the PR, I was going to ask whether Location and Content-Location can be trailers, but found that this is consistent with RFC 9110.

Will have a look at the full text later to see how in integrates, and whether we may want to have more examples.

**Comment by reschke** [comment]
OK, I think we should add an example to section 4.

@MikeBishop - can you move this PR to the WG repo, so I can add to it?

**Comment by MikeBishop** [comment]
@reschke, GitHub doesn't let me change the source branch of a PR, only the target. I've pushed the same branch to the upstream repo and opened #2851 with the same commit history. You should have full access to that one.

**Pull request closed without merging**